### PR TITLE
New TAngularCorrelation Updates

### PR DESCRIPTION
**Pull request by SmithJK**
Including grouping and folding.

### CODE REVIEW

File: include/TAngularCorrelation.h [hunk 1]
```diff
@@ -59,15 +79,16 @@ class TAngularCorrelation : public TObject {
...
       TH1D* IntegralSlices(TH2* hst, Double_t min, Double_t max);
       TH1D* FitSlices(TH2* hst,TPeak* peak,Bool_t visualization);
       TH1D* DivideByWeights(TH1* hst);
       void DivideByWeights();
       void UpdatePeak(Int_t index,TPeak* peak);
-      TGraphAsymmErrors* CreateGraphFromHst(TH1* hst);
-      TGraphAsymmErrors* CreateGraphFromHst() { return CreateGraphFromHst(fIndexCorrelation); }
+      void ScaleSingleIndex(TH1* hst, Int_t index, Int_t factor);
+      TGraphAsymmErrors* CreateGraphFromHst(TH1* hst, Bool_t fold, Bool_t group);
+    //  TGraphAsymmErrors* CreateGraphFromHst() { return CreateGraphFromHst(fIndexCorrelation); }
```
Comment (SmithJK) [inline include/TAngularCorrelation.h]: Why did you remove this function?

File: include/TAngularCorrelation.h [hunk 2]
```diff
@@ -77,14 +98,29 @@ class TAngularCorrelation : public TObject {
...
+      void PrintFoldedGroupIndexes(); // print the map
+      
       //Int_t SetAngleMap(Double_t* angles); // sets the angles in the map, with an array of angles, where the angular index is determined by the index of the array element
       //Int_t SetWeights(Int_t* weights); // input is weight array itself
       static std::vector<Double_t> GenerateAngleMap(std::vector<Int_t> &arraynumbers, std::vector<Int_t> &distances);
       static std::map<Int_t,std::map<Int_t,Int_t>> GenerateIndexMap(std::vector<Int_t> &arraynumbers, std::vector<Int_t> &distances, std::vector<Double_t> &anglemap);
       static std::vector<Int_t> GenerateWeights(std::vector<Int_t> &arraynumbers, std::vector<Int_t> &distances, std::map<Int_t,std::map<Int_t,Int_t>> &indexmap); // with input of array number array (crystals that were present in data collection), generates the weights for each angular index (no input generates weights for 16 detectors)
+static std::vector<Int_t> GenerateGroupWeights(std::vector<Int_t> &group, std::vector<Int_t> &weights);
+static std::vector<Int_t> GenerateFoldedWeights(std::vector<Int_t> &folds, std::vector<Int_t> &weights);
```
Comment (SmithJK) [inline include/TAngularCorrelation.h]: Indention.

File: include/TAngularCorrelation.h [hunk 3]
```diff
@@ -49,7 +59,17 @@ class TAngularCorrelation : public TObject {
       Double_t GetAngleFromIndex(Int_t index) { return fAngleMap[index]; } // returns the opening angle for a specific angular index
       //TODO: move the next function to implementation file and check if in range
       Double_t GetWeightFromIndex(Int_t index) { return fWeights[index]; } // returns the weight for a specific angular index
+      Double_t GetGroupFromIndex(Int_t index) { return fGroups[index]; } // returns the assigned group for a specific angular index
+      Double_t GetGroupWeightFromIndex(Int_t index) { return fGroupWeights[index]; }//returns the determined weight for each group
+      Double_t GetGroupAngleFromIndex(Int_t index) { return fGroupAngles[index]; } // returns the average angle for each group
+      Double_t GetFoldedAngleFromIndex(Int_t index) { return fFoldedAngles[index]; } // returns the folded angle value for each angular index
+      Double_t GetFoldedAngularIndex(Int_t index) { return fFoldedAngularIndexes[index]; } // returns the folded index for each angular index
+      Double_t GetFoldedAngleWeightsFromIndex(Int_t index) { return fFoldedAngularWeights[index]; } //returns the weight for each folded angle
```
Comment (SmithJK) [inline include/TAngularCorrelation.h]: Weights should probably be singular - same below.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 4]
```diff
@@ -55,24 +55,24 @@ TH2D* TAngularCorrelation::Create2DSlice(THnSparse *hst, Double_t min, Double_t
    Int_t indexaxis,energy1axis,energy2axis;
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:57]: Note for JKS: there should probably be a check to make sure that the THnSparse is of three-dimensions first.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 5]
```diff
@@ -399,17 +491,54 @@ TH1D* TAngularCorrelation::FitSlices(TH2* hst, TPeak* peak, Bool_t visualization
 /// \param[in] hst One-dimensional histogram of angular index vs. counts
 ///
 
-TGraphAsymmErrors* TAngularCorrelation::CreateGraphFromHst(TH1* hst)
+TGraphAsymmErrors* TAngularCorrelation::CreateGraphFromHst(TH1* hst, Bool_t fold, Bool_t group)
 {
    TGraphAsymmErrors* graph = new TGraphAsymmErrors();
-   
+if(!fold && !group){
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: formatting (space needed at beginning of lines)

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 6]
```diff
@@ -418,7 +547,8 @@ TGraphAsymmErrors* TAngularCorrelation::CreateGraphFromHst(TH1* hst)
 
       // fill graph with point
       Int_t graphn = graph->GetN();
-      graph->SetPoint(graphn,TMath::Cos(angle),y);
+      if(!fold) graph->SetPoint(graphn,TMath::Cos(angle),y);
+      else if(fold) graph->SetPoint(graphn,angle,y);
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: Are the elements in the folded angles array actually the cosine values already? Otherwise, I don't see a reason this needs to have if statements in it.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 7]
```diff
@@ -478,6 +609,11 @@ Bool_t TAngularCorrelation::CheckMaps()
 void TAngularCorrelation::PrintIndexMap()
 {
    Int_t size = fIndexMapSize;
+     if(size == 0){
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: I like this addition.

Check the begin-of-line spacing.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 8]
```diff
@@ -522,6 +658,63 @@ void TAngularCorrelation::PrintWeights()
...
+   for (Int_t i=0;i<size;i++) {
+      printf("||  %-13i  |  %-19.4f  |  %-6.0f  ||\n",i,GetAngleFromIndex(i),GetGroupFromIndex(i));
+   }
+   printf("--------------------------------------------------------\n");
+
+   return;
+}
+////////////////////////////////////////////////////////////////////////////////
+/// Prints map of group vs. group weight vs. average group angle
+///
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:883]: Are these average group angles calculated in the code, or input by the user?

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 9]
```diff
@@ -530,6 +723,11 @@ void TAngularCorrelation::PrintWeights()
 void TAngularCorrelation::PrintAngleMap()
 {
    Int_t size = fAngleMap.size();
+     if(size == 0){
+      printf("Folded angles have not been assigned yet.\n");
+      printf("Therefore cannot print map.\n");
+      return;
+      }
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:921]: Again, _thumbs up_ for this addition, and check the spacing.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 10]
```diff
@@ -541,6 +739,79 @@ void TAngularCorrelation::PrintAngleMap()
...
+void TAngularCorrelation::PrintFoldedAngleMap()
+{
+   Int_t size = fFoldedAngles.size();
+
+   if(size == 0){
+      printf("Folded angles have not been assigned yet.\n");
+      printf("Therefore cannot print map.\n");
+      return;
+      }
+
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: Spacing, above and below here.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 11]
```diff
@@ -642,7 +913,95 @@ std::vector<Int_t> TAngularCorrelation::GenerateWeights(std::vector<Int_t> &arra
 
    return weights;
 }
+////////////////////////////////////////////////////////////////////////////////
+/// Creates map of group weights vs. angular index
+///
+/// \param[in] arraynumbers Vector of array numbers used in this experiment
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: These inputs need to be updated.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 12]
```diff
@@ -642,7 +913,95 @@ std::vector<Int_t> TAngularCorrelation::GenerateWeights(std::vector<Int_t> &arra
...
+      groupweights.push_back(groupArray[i]);
+   }
+   
+  
+  return groupweights;
+}
+////////////////////////////////////////////////////////////////////////////////
+/// Creates map of folded weights vs. folds
+///
+/// \param[in]
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: ?

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 13]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+std::vector<Int_t> TAngularCorrelation::AssignGroups(std::vector<Int_t> &group, std::vector<Double_t> &anglemap)
+{
+   std::vector<Int_t> groups; // vector to return
+
+   // get number of group entries
+  Int_t size = group.size();
+
+// basic consistency check
+  if (size!=(Int_t) anglemap.size()) {     
+ }
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: Probably want something in these brackets.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 14]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+  if (size!=(Int_t) anglemap.size()) {     
+ } 
+   
+   // initialize group vector
+   for (Int_t i=0;i<size;i++) {
+      groups.push_back(group[i]);
+   }
+
+   return groups;
+}
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: Also need to assign groups to fGroups so it stays available.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 15]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+////////////////////////////////////////
+/// Assign average angles for groups
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: I think "average" is the wrong word to use here - I think "Assign angles for groups" will work just fine.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 16]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+      printf("Not all groups have been assigned an angle.\n");   
+   } 
+   
+   // initialize group vector
+   for(Int_t i=0;i<size;i++) {
+       groupangle.push_back(groupangles[i]);
+   }
+
+   return groupangle;
+}
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: groupangle also needs to be assigned to a class member so we can keep it around.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 17]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+////////////////////////////////////////
+/// Generate Folded Angles
+///gives Cos(RADIAN) angles out for each folded index
+///
+/// \param[in] anglemap (can be for grouped or ungrouped anglular indexes)
+
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: We really should be consistent here - since this is an angle array, it should give the angles, rather than the cosine of them.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 18]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+  }
+  
+  
+  std::sort(folds.begin(),folds.end()); 
+
+   return folds;
+}
 
+////////////////////////////////////////
+/// Assign Groups for Angular Indexes
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: This documentation needs to be updated - probably a relic of copy-paste.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 19]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+         Double_t fold_angle = folds[j];
+         if(TMath::Abs(fold_angle-cos_angle)<0.00005){//compare abs(cos(angle)) to fold_Angles to find matches
+            fold_indexes.push_back(j);//assign folds 
+            }
+         }
+     }
+ 
+
+   return fold_indexes;
+}
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:1397]: Since GenerateFoldedAngles and GenerateFoldedIndexes are doing very similar things, can we combine the two into a single function?

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 20]
```diff
@@ -697,7 +1056,146 @@ std::map<Int_t,std::map<Int_t,Int_t>> TAngularCorrelation::GenerateIndexMap(std:
...
+       } 
+     
+            
+  }
+  
+  
+  std::sort(folds.begin(),folds.end()); 
+
+   return folds;
+}
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: folds needs to be assigned to the class.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 6 above)
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: I found the answer later on - I suggest we actually save the angle values in radians and eliminate this if statement.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 21]
```diff
@@ -729,9 +1227,52 @@ Int_t TAngularCorrelation::GenerateMaps(std::vector<Int_t> &arraynumbers, std::v
    fNumIndices = fAngleMap.size();
    fIndexMap = GenerateIndexMap(arraynumbers,distances,fAngleMap);
    fWeights = GenerateWeights(arraynumbers,distances,fIndexMap);
+   fFoldedAngles = GenerateFoldedAngles(fAngleMap);
+   fFoldedAngularIndexes = GenerateFoldedIndexes(fFoldedAngles, fAngleMap);
+   fFoldedAngularWeights = GenerateFoldedWeights(fFoldedAngularIndexes, fWeights);
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: So this function doesn't need the group specification as input? Because it doesn't look like it's being called here. If it *does* need the group specification, I suggest we modify this somehow so someone could do an AC with just the original indices and not need to specify the groups. I have no objection to calculating the folded weights here though.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 22]
```diff
@@ -794,7 +1335,6 @@ Int_t TAngularCorrelation::GenerateMaps(Int_t detectors, Int_t distance)
    return val;
 }
 
-////////////////////////////////////////////////////////////////////////////////
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:797]: Why was this line dropped? I think it's necessary for the doxygen documentation.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 23]
```diff
@@ -935,7 +1496,7 @@ TH1D* TAngularCorrelation::DivideByWeights(TH1* hst)
       Int_t index = hst->GetBinLowEdge(i);
       Double_t content = hst->GetBinContent(i);
       Double_t error = hst->GetBinError(i);
-      Double_t weight = this->GetWeightFromIndex(index);
+      Double_t weight = this->GetGroupWeightFromIndex(index);
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: This will screw up using DivideByWeights for a non-grouped spectrum.

File: include/TAngularCorrelation.h (hunk 1 above)
Comment (jpore) [inline include/TAngularCorrelation.h]: doesn't need to be removed

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 8 above)
Comment (jpore) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:883]: these values are input by the user.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 19 above)
Comment (jpore) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:1397]: I put them as two separate functions because they are making two separate vectors that we need. It is kinda like the index map vs. the angle map.
1. the GenerateFoldedAngles() makes a vector that is the folded index versus the folded angle. This vector is as long as how many folds there are. We need this for assigning the folds in the create2Dslice.
2. The GenerateFoldedIndexes() makes a vector that is the angular index versus the folded index. This vector is as long as the number of angular indexes. 

I am not sure how to combine them into one function and still get out both of the vectors that we need. Let me know if you see a way, it would be great to do so.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 22 above)
Comment (jpore) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:797]: hmm, I am not sure what happened to this line. I have put it back.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 23 above)
Comment (jpore) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: this is now fixed, and I have added grouped and folded weights in as well.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 6 above)
Comment (jpore) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: I have resolved this issue. Now the angles are in radians.

Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: I think you fixed the function output, but not this if statement.

**Comment by SmithJK** [comment]
Please replace "average" with "assigned" everywhere; it's more precise and accurate.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 19 above)
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:1397]: On a second (and third) evaluation, I can't see a way around it. Let's keep it this way.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 6 above)
Comment (jpore) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: I already noticed this and have fixed it.

----- Original Message -----

**Comment by SmithJK** [comment]
Can we create a "Modify2DSlice" function to do all the folding and grouping? This would allow us to use the same folding and grouping function regardless of the way the 3D matrices are stored (as an array of TH2, an array of THnSparse, or a THnSparse 3D matrix).

input: original angular index vs. energy 2D histogram, folded boolean, grouped boolean
output: modified index vs. energy 2D histogram

**Comment by jpore** [comment]
yes, I do think that we can create such a function. Do you want me to work on this?

----- Original Message -----

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx (hunk 6 above)
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx]: I'll watch for the push of that modification then.

**Comment by SmithJK** [comment]
Yes, please include it.

File: libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx [hunk 24]
```diff
@@ -642,6 +975,49 @@ std::vector<Int_t> TAngularCorrelation::GenerateWeights(std::vector<Int_t> &arra
...
+    }
+  
+   // initialize vector
+   for (Int_t i=0;i<=max;i++) {
+      modifiedweights.push_back(tempArray[i]);
+   }
+   
+  
+  return modifiedweights;
+}
```
Comment (SmithJK) [inline libraries/TGRSIAnalysis/TAngularCorrelation/TAngularCorrelation.cxx:1183]: I like the addition of this function.
